src/backend/impl/src/services/proposal_review_service.rs

```diff
@@ -0,0 +1,1206 @@
+        self.validate_fields(request.summary.as_ref(), request.review_duration_mins)?;
+
+        let user_id = self
+            .user_profile_repository
+            .get_user_id_by_principal(&calling_principal)
+            .ok_or_else(|| {
+                ApiError::not_found(&format!(
+                    "User id for principal {} not found",
+                    calling_principal.to_text()
+                ))
+            })?;
+
+        let proposal_id = Uuid::try_from(request.proposal_id.as_str())?;
+
+        match self.proposal_repository.get_proposal_by_id(&proposal_id) {
+            Some(proposal) => {
+                if proposal.is_completed() {
+                    return Err(ApiError::conflict(&format!(
+                        "Proposal with Id {} is already completed",
+                        proposal_id.to_string()
+                    )));
+                }
+            }
+            None => {
+                return Err(ApiError::not_found(&format!(
+                    "Proposal with Id {} not found",
+                    request.proposal_id
+                )))
+            }
+        }
+
+        let date_time = get_date_time()?;
+
+        let proposal_review = ProposalReview {
+            proposal_id,
+            user_id,
+            status: ProposalReviewStatus::Draft,
+            created_at: DateTime::new(date_time)?,
+            summary: request.summary.unwrap_or("".to_string()),
+            review_duration_mins: request.review_duration_mins.unwrap_or(0),
+            build_reproduced: request.build_reproduced.unwrap_or(false),
+            // TODO: use reproduced_build_image_id from request
+            reproduced_build_image_id: None,
+        };
+
+        let id = self
+            .proposal_review_repository
+            .create_proposal_review(proposal_review.clone())
+            .await?;
+
+        Ok(map_create_proposal_review_response(id, proposal_review))
+    }
+
+    fn update_proposal_review(
+        &self,
+        calling_principal: Principal,
+        request: UpdateProposalReviewRequest,
+    ) -> Result<(), ApiError> {
+        self.validate_fields(request.summary.as_ref(), request.review_duration_mins)?;
+
+        let user_id = self
+            .user_profile_repository
+            .get_user_id_by_principal(&calling_principal)
+            .ok_or_else(|| {
+                ApiError::not_found(&format!(
+                    "User id for principal {} not found",
+                    calling_principal.to_text()
+                ))
+            })?;
+
+        let id = Uuid::try_from(request.id.as_str())?;
+        let mut current_proposal_review = self
+            .proposal_review_repository
+            .get_proposal_review_by_id(&id)
+            .ok_or_else(|| {
+                ApiError::not_found(&format!("Proposal review with Id {} not found", request.id))
+            })?;
+
+        if current_proposal_review.user_id != user_id {
```

Comment:
What if the parameters was proposal Id instead of proposal review Id? Then the API would just edit the proposal belonging to the requesting principal for this proposal, if it exists. Then it's not possible for someone to even try edit someone else's review. This would also be then more consistent with fetching reviews by user Id, or individual review by proposal Id and user Id.

Reply:
Yeah, this makes much more sense